### DISCUSSION TITLE
Delimited multi-shot continuations on Java Loom

### DISCUSSION THREAD

**Discussion by steshaw** [javactrl/javactrl#6] [General]
I just came across [this implementation of delimited continuations in Java on top of Loom](https://gist.github.com/b-studios/65e9e614f006d2125659267f6113efac) and wondered what your opinion is.

**Comment by awto** [comment]
Loom only supports single-shot continuations, and there are no plans to add multi-shot. This makes it unsuitable for tasks like stream processing, or logic or probabilistic programming. Additionally, Loom doesn't yet support serializable continuations, though this is something the developers plan to add in the future.

I would be thrilled if Loom eventually supports multi-shot continuations. It would make Loom a much simpler alternative to using instrumentation like in javactrl. Perhaps something will be possible once serialization is added.

**Reply by steshaw** [reply]
The Gist I referred to apparently implements multi-shot (aka multiprompt) continuations on top of Loom! This is why I wanted to point it out to you and get your opinion on it.

**Reply by awto** [reply]
Multi-shot and multi-prompt refer to different concepts. Multiprompt generally means having several independent boundaries, whereas multi-shot usually refers to the same continuation being executed multiple times. The Gist mentions it's a one-shot on line 43. There are also no examples using the same continuation more than once.

**Reply by steshaw** [reply]
Oops, sorry to bother you with that, then! I had suspected that @b-studios had come up with a way of having full, multi-shot, delimited continuations on the JVM for the [Effekt](https://github.com/effekt-lang/effekt) programming language with effects. I got a bit excited when I saw that Gist, as didn't realise that multiprompt was something else altogether.

The only other related thing I came across in my search was an [implementation of multi-shot continuations in Kotlin for Arrow](https://github.com/arrow-kt/arrow-core/pull/226), built on top of coroutines. It's limited in some way that I don't yet understand. Also, I'm not sure where this code has ended up after getting merged into https://github.com/arrow-kt/arrow/ and seemingly moved around since 2020.

**Reply by b-studios** [reply]
Hey, a few hopefully clarifying words:

Effekt-the-programming-language is different from our earlier experiments as libraries (in [Scala](https://github.com/b-studios/scala-effekt) and [Java](https://github.com/b-studios/java-effekt)).

- **Effekt**: a standalone programming language that compiles to JS, LLVM, and many other backends. For these backends we typically have our own runtime system that supports multishot.
- **Scala-Effekt**: we use a monad that supports multishot and users need to program in monadic style
- **Java-Effekt**: we instrument the JVM bytecode to avoid users having to program in monadic style

Both Scala-Effekt and Java-Effekt are not longer maintained as all our efforts are on the Effekt programming language.

The crucial problem with multishot on Loom / Java is that you would need to copy the continuation in a proper way. Honestly I don't know what the current status is, but it hasn't been possible in the past.

**Reply by steshaw** [reply]
Thanks for chiming in, @b-studios!
